File: problems/16.graphs/ts/cal-minutes.ts
export function calculateNumberOfMinutes(
    managers: number[],
    informTime: number[]
) {
    const headId = managers.findIndex((manager) => manager === -1);
    let result = 0;
    const queue = [headId];
    let queueLength = queue.length;
    let times = [];
    const seen = new Map<number, boolean>();
    seen.set(headId, true);
    result += informTime[headId];

    while (queue.length > 0) {
        const vertex = queue.shift() as number;
        queueLength -= 1;
        let connections = [];
        managers[vertex] !== -1 &&
            managers[vertex] !== vertex &&
            connections.push(managers[vertex]);
        connections = [
            ...connections,
            ...managers
                .map((manager, index) => (manager === vertex ? index : -1))
                .filter((item) => item !== -1),
        ];

        for (let i = 0; i < connections.length; i++) {
            const connection = connections[i];
            if (!seen.get(connection) && informTime[connection]) {
                times.push(informTime[connection]);
                queue.push(connection);
            }
            seen.set(connection, true);
        }

        if (queueLength === 0) {
            const timeForLevel =
                times.length > 0 ? Math.max.apply(null, [...times]) : 0;
            times = [];
            result += timeForLevel;
            queueLength = queue.length;
        }
    }

    return result;
}

File: problems/16.graphs/golang/main/cal_minutes.go
package main

// solving problem with bfs
func calculateMinutes1(managers, minutes []int, headId int) int {
	delay := 0
	adjList := make([][]int, len(managers))

	for employeeId, manager := range managers {
		if manager != -1 {
			adjList[manager] = append(adjList[manager], employeeId)
		}
	}

	queue := []int{headId}
	seen := make(map[int]bool, len(managers))
	seen[4] = true

	for len(queue) > 0 {
		vertex := queue[len(queue)-1]
		queue = queue[:len(queue)-1]
		connections := adjList[vertex]

		for _, connection := range connections {
			if !seen[connection] {
				delay += minutes[connection]
				seen[connection] = true
				queue = append(queue, connection)
			}
		}
	}

	return delay
}

func dfs(minutes []int, vertex int, adjList [][]int, seen map[int]bool, delay int) int {
	seen[vertex] = true
	delay += minutes[vertex]
	connections := adjList[vertex]

	for _, connection := range connections {
		if !seen[connection] {
			return dfs(minutes, connection, adjList, seen, delay)
		}
	}

	return delay
}

// solving problem with dfs
func calculateMinutes2(managers, minutes []int, headId int) int {
	adjList := make([][]int, len(managers))
	seen := make(map[int]bool, len(managers))

	for employeeId, manager := range managers {
		if manager != -1 {
			adjList[manager] = append(adjList[manager], employeeId)
		}
	}

	return dfs(minutes, headId, adjList, seen, 0)
}
